docs/user-guide/05-compiling.md

```diff
@@ -0,0 +1,493 @@
+
+Applications that more control over compilation are encouraged to use the C++ compilation API described in the next section.
+
+Using the Compilation API
+-------------------------
+
+The C++ API provided by Slang is meant to provide more complete control over compilation for applications that need it.
+The additional level of control means that some tasks require more individual steps than they would when using a one-size-fits-all tool like `slangc`.
+
+### "COM-lite" Components
+
+Many parts of the Slang C++ API use interfaces that follow the design of COM (the Component Object Model).
+Some key Slang interfaces are binary-compatible with existing COM interfaces.
+However, the Slang API does not depend on any runtime aspects of the COM system, even on Windows; the Slang system can be seen as a "COM-lite" API.
+
+The `ISlangUnknown` interface is equivalent to (and binary-compatible with) the standard COM `IUnknown`.
+Application code is expected to correctly maintain the reference counts of `ISlangUnknown` objects returned from API calls; the `SlangComPtr<T>` "smart pointer" type is provided as an optional convenience for applications that want to use it.
+
+Many Slang API calls return `SlangResult` values; this type is equivalent to (and binary-compatible with) the standard COM `HRESULT` type.
+As a matter of convention, Slang API calls return a zero value (`SLANG_OK`) on success, and a negative value on errors.
+
+### Creating a Global Session
+
+A Slang _global session_ uses the interface `slang::IGlobalSession` and it represents a connection from an application to a particular implementation of the Slang API.
+A global session is created using the function `slang::createGlobalSession()`:
+
+```c++
+SlangComPtr<IGlobalSession> globalSession;
+slang::createGlobalSession(globalSession.writeRef());
+```
+
+When a global session is created, the Slang system will load its internal representation of the _standard library_ that the compiler provides to user code.
+The standard library can take a significant amount of time to load, so applications are advised to use a single global session if possible, rather than creating and then disposing of one for each compile.
+
+> #### Note ####
+> Currently, the global session type is *not* thread-safe.
+> Applications that wish to compile on multiple threads will need to ensure that each concurrent thread copmiles with a distinct global session.
+
+### Creating a Session
+
+A _session_ uses the interface `slang::ISession`, and represents a scope for compilation with a consistent set of compiler options.
+In particular, all compilation with a single session will share:
+
+* A list of enabled compilation targets (with their options)
+
+* A list of search paths (for `#include` and `import`)
+
+* A list of pre-defined macros
+
+In addition, a session provides a scope for the loading and re-use of modules.
```

Comment:
It is the kind of thing we *could* mention, but I suspect that is the kind of thing where users wouldn't even think to ask about it if we didn't bring it up as a question.

I suspect that a lot of the editing passes we will need to make on the guide material will actually be to *remove* details like that rather than to add them, because we should be aiming to minimize the number of hoops users jump through to get up and running. Once somebody is using the system and wants to start caring about the details, more detailed documentation will be helpful (and that is where doxygen-generated documentation for our C++ API is probably useful to have).